lib/acme/client.rb

```diff
@@ -127,8 +128,16 @@ def finalize(url:, csr:)
     Acme::Client::Resources::Order.new(self, **arguments)
   end
 
-  def certificate(url:)
+  def certificate(url:, fetch_alternative_chains: false)
     response = download(url, format: :pem)
+    if fetch_alternative_chains
+      @alternative_certificates = []
+      alt_chains_urls = fetch_alternative_links(response)
+      alt_chains_urls.each do |alt_chains_url|
+        response = download(alt_chains_url, format: :pem)
+        @alternative_certificates.push response.body
+      end
+    end
     response.body
   end
```

Comment:
I think it might be nicer to have a separate method entirely just for `get_all_certificates` or something. I think that storing a single response data on the client object itself makes it harder to explain/trace through whats going on. A new method returning an array wouldn't need to use state to avoid changing the existing methods like this

Reply:
https://github.com/unixcharles/acme-client/pull/187#discussion_r478555219

As I said in the comment above, I'll try to set the `LInks` in an instance valiable, so I'll separate the methods accordingly. Thanks 😃

Reply:
> I think it might be nicer to have a separate method entirely just for get_all_certificates or something. I think that storing a single response data on the client object itself makes it harder to explain/trace through whats going on. A new method returning an array wouldn't need to use state to avoid changing the existing methods like this

I agree.

`#certificate(url: )` should work as expected assuming a URL pointing to an alternate. It's a bit of a weird return value.

You would end up with

```
_, alternatives = client.certificate(url: url)
certificate = client.certificate(url: alternative.first)
```

I would rather not change the existing API of the client. Especially method signature.

Could we go with something like

```
> client.alternative_certificates(url: certificate_url)
=> ["alternative_certificate_pem", ...]
> order. alternative_certificates
=> ["alternative_certificate_pem", ...]
```

Do you guys have a use case for fetching the certificate and the alternative from one method except for the fact that you already have it available? I'm assuming you would want one or the other.

Reply:
I think the implementation is going to be:
- fetch the main cert and check root
- fetch alternatives looking for the wanted root

Like mentioned in another place I think the certbot API is nicer if we're okay using the x509 stdlib here to do some cert parsing to check the root cert common-name, instead of managing the main/alternative choice manually. That lets us ask the client to fetch us the IdenTrust certificate and it will automatically handle the switchover from main-cert to alternative-cert whenever LE swaps those two.

Reply:
hmm...

for example, Should I implement the following?


```
$ order.certificate
#=> returns default certificate


$ order.certificate(preferred_chain: 'DST Root CA X3')  # option name come from certbot.
#=> returns 'DST Root CA X3' certificate

$ order.certificate(preferred_chain: 'not available chain name')
#=> raise error
```

Reply:
By the way, do you know a good way to compare the common name of the `OpenSSL::X509::Name` object?

```
> cert.issuer
=> #<OpenSSL::X509::Name CN=cackling cryptographer fake ROOT>


> cert.issuer.to_s
=> "/CN=cackling cryptographer fake ROOT"

> cert.issuer.to_a
=> [["CN", "cackling cryptographer fake ROOT", 12]]
```

It seems to me that #to_a is sufficient for now, but it's somewhat inaccurate.

Reply:
```
> OpenSSL::X509::Certificate.new(xx)
=> #<OpenSSL::X509::Certificate
 subject=#<OpenSSL::X509::Name CN=Let's Encrypt Authority X3,O=Let's Encrypt,C=US>,
 issuer=#<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>,
 serial=#<OpenSSL::BN 13298795840390663119752826058995181320>,
 not_before=2016-03-17 16:40:46 UTC,
 not_after=2021-03-17 16:40:46 UTC>

> OpenSSL::X509::Certificate.new(xx).issuer.to_a
=> [["O", "Digital Signature Trust Co.", 19], ["CN", "DST Root CA X3", 19]]

> OpenSSL::X509::Certificate.new(xx).issuer.to_a.flatten
=> ["O", "Digital Signature Trust Co.", 19, "CN", "DST Root CA X3", 19]
```

hmm...

Reply:
@ryansouza @unixcharles 
I've chaned an option to preferred_chain so that it will always return a single certificate chain. What do you think?

Reply:
For me the comment still stand. `Acme::Client#certificate` returning is no a consistent `API`

`Acme::Client#order` return a order, `Acme::Client#order` return an account, `Acme::Client#authorization` return an authorization.

Now, `#certificate` would be returning an array of `[Certificate, [links...]]`, which would break the API and require a major release and make it less consistent. I would not expect a method called `#certificate` to have this type of return value.

Reply:
@unixcharles Thanks for comment 👍  I understand your opinion that we should avoid changing the API. 

So which of the following do you think is better?

- Add the preferred_chain option to `Acme::Client#certificate` to change it to return an alternate certificate directly
- Add the Acme::Client#altenate_certificate(preferred_chain: foobar) method and use it to get an alternate_certificate